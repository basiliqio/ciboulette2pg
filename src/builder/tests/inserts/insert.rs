use super::*;

#[test]
fn simple() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_external_single() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"articles":
				{
					"data":
					{
						"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
						"type": "articles"
					}
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_external_multiple() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"articles":
				{
					"data":
					[
						{
							"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
							"type": "articles"
						},
						{
							"id": "5043c5ef-debd-4b60-b905-7dab2a7338bc",
							"type": "articles"
						}
					]
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_external_multiple_include() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples?include=articles").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"articles":
				{
					"data":
					[
						{
							"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
							"type": "articles"
						},
						{
							"id": "5043c5ef-debd-4b60-b905-7dab2a7338bc",
							"type": "articles"
						}
					]
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_external_multiple_include_rel() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url =
        Url::parse("http://localhost/peoples?include=favorite_color,articles,people-article")
            .unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"articles":
				{
					"data":
					[
						{
							"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
							"type": "articles"
						},
						{
							"id": "5043c5ef-debd-4b60-b905-7dab2a7338bc",
							"type": "articles"
						}
					]
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_internal_single() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"favorite_color":
				{
					"data":
					{
						"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
						"type": "favorite_color"
					}
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_internal_single_include() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url = Url::parse("http://localhost/peoples?include=favorite_color").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"favorite_color":
				{
					"data":
					{
						"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
						"type": "favorite_color"
					}
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}

#[test]
fn relationship_internal_single_include_all() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store();
    let parsed_url =
        Url::parse("http://localhost/peoples?include=favorite_color,articles,people-article")
            .unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Create;
    const BODY: Option<&str> = Some(
        r#"
	{
		"data":
		{
			"id": "6720877a-e27e-4e9e-9ac0-3fff4deb55f2",
			"type": "peoples",
			"attributes":
			{
				"first-name": "Hello",
				"last-name": "World"
			},
			"relationships":
			{
				"favorite_color":
				{
					"data":
					{
						"id": "8759b021-0932-40cb-8386-e1a83da1c73e",
						"type": "favorite_color"
					}
				}
			}
		}
	}
	"#,
    );

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &BODY);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteCreateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_insert(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let res = builder.build().unwrap();

    insta::assert_debug_snapshot!(res);
}
