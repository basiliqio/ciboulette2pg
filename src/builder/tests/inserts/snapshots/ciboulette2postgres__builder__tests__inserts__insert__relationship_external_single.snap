---
source: src/builder/tests/inserts/insert.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
WITH "cte_peoples_insert" AS (
  INSERT INTO
    "public"."peoples" ("first-name", "last-name")
  VALUES
    ($0, $1) RETURNING *
),
"cte_peoples_data" AS (
  SELECT
    "cte_peoples_insert"."id" :: TEXT AS "id",
    $2 :: TEXT AS "type",
    JSON_BUILD_OBJECT(
      $3,
      "cte_peoples_insert"."age",
      $4,
      "cte_peoples_insert"."first-name",
      $5,
      "cte_peoples_insert"."gender",
      $6,
      "cte_peoples_insert"."last-name",
      $7,
      "cte_peoples_insert"."twitter"
    ) AS "data"
  FROM
    "cte_peoples_insert"
),
"cte_favorite_color_data" AS (
  SELECT
    "public"."favorite_color"."id" :: TEXT AS "id",
    $8 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."favorite_color"
  WHERE
    "public"."favorite_color"."id" IN (
      SELECT
        "cte_peoples_insert"."favorite_color"
      FROM
        "cte_peoples_insert"
    )
),
"cte_rel_articles_id" AS (
  VALUES
    ($9 :: uuid)
),
"cte_rel_articles_insert" AS (
  INSERT INTO
    "public"."people-article" ("article_id", "people_id")
  SELECT
    "cte_peoples_data"."id" AS "article_id",
    "cte_rel_articles_id"."id" AS "people_id"
  FROM
    "cte_peoples_data",
    "cte_rel_articles_id" RETURNING "public"."people-article"."id",
    "public"."people-article"."article_id",
    "public"."people-article"."people_id"
),
"cte_rel_articles_rel_data" AS (
  SELECT
    "public"."people-article"."id" :: TEXT AS "id",
    $10 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."people-article"
  WHERE
    "public"."people-article"."people_id" = "cte_peoples_data"."id"
),
"cte_rel_articles_data" AS (
  SELECT
    "public"."articles"."id" :: TEXT AS "id",
    $11 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."articles"
  WHERE
    "public"."articles"."id" IN (
      SELECT
        "cte_rel_articles_rel_data"."article_id"
      FROM
        "cte_rel_articles_rel_data"
    )
)
SELECT
  "cte_rel_articles_data"."id",
  "cte_rel_articles_data"."type",
  "cte_rel_articles_data"."data"
FROM
  "cte_rel_articles_data"
UNION
ALL
SELECT
  "cte_favorite_color_data"."id",
  "cte_favorite_color_data"."type",
  "cte_favorite_color_data"."data"
FROM
  "cte_favorite_color_data"
UNION
ALL
SELECT
  "cte_rel_articles_rel_data"."id",
  "cte_rel_articles_rel_data"."type",
  "cte_rel_articles_rel_data"."data"
FROM
  "cte_rel_articles_rel_data"
UNION
ALL
SELECT
  "cte_peoples_data"."id",
  "cte_peoples_data"."type",
  "cte_peoples_data"."data"
FROM
  "cte_peoples_data";
